### PR TITLE
Expose cudaStreamCaptureMode in CUDA Graphs, use local setting in inductor

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107407

>  capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream. 
Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
 may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
 actions in the current thread, and "relaxed" will not error on these actions.

Inductor codegen is single-threaded, so it should be safe to enable "thread_local" for inductor's cuda graph capturing. We have seen errors when inductor cudagraphs has been used concurrently with data preprocessing in other threads.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

Differential Revision: [D48656014](https://our.internmc.facebook.com/intern/diff/D48656014)

### CODE REVIEW

File: c10/cuda/CUDACachingAllocator.cpp [hunk 1]
```diff
@@ -2808,6 +2808,7 @@ class DeviceCachingAllocator {
 
   void release_block(Block* block) {
     TORCH_INTERNAL_ASSERT(!block->expandable_segment_);
+    std::cout << " Freeing tensor \n";
```
Comment (albanD) [inline c10/cuda/CUDACachingAllocator.cpp:2811]: Debug print

File: torch/csrc/cuda/Graph.cpp [hunk 2]
```diff
@@ -30,9 +31,30 @@ void THCPGraph_init(PyObject* module) {
...
+                ? pool_opt.value()
+                : c10::cuda::MempoolId_t{0, 0};
+            if (capture_error_mode == "global") {
+              capture_mode = cudaStreamCaptureModeGlobal;
+            } else if (capture_error_mode == "thread_local") {
+              capture_mode = cudaStreamCaptureModeThreadLocal;
+            } else if (capture_error_mode == "relaxed") {
+              capture_mode = cudaStreamCaptureModeRelaxed;
+            } else {
+              TORCH_INTERNAL_ASSERT(
```
Comment (albanD) [inline torch/csrc/cuda/Graph.cpp:48]: TORCH_CHECK Since this is a user facing API?

File: torch/csrc/cuda/Graph.cpp [hunk 3]
```diff
@@ -30,9 +31,30 @@ void THCPGraph_init(PyObject* module) {
...
+                  false,
+                  "Unknown capture error mode. Expected `global`, `thread_local`, or `relaxed`, got ",
+                  capture_error_mode);
+            }
+            return self.capture_begin(pool, capture_mode);
+          },
+          py::arg("pool") = (c10::optional<c10::cuda::MempoolId_t>)c10::nullopt,
```
Comment (albanD) [inline torch/csrc/cuda/Graph.cpp:55]: Why was this changed to a nullopt?
Also these args are always passed so no need to provide default for them?

File: torch/csrc/cuda/Graph.cpp [hunk 4]
```diff
@@ -30,9 +31,30 @@ void THCPGraph_init(PyObject* module) {
...
+          py::arg("capture_error_mode") = "global",
+          py::call_guard<py::gil_scoped_release>())
```
Comment (albanD) [inline torch/csrc/cuda/Graph.cpp:57]: Are you sure this is safe?
This should be quite fast anyways right?

File: torch/cuda/graphs.py [hunk 5]
```diff
@@ -68,13 +68,13 @@ def capture_begin(self, pool=None):
             pool (optional): Token (returned by :func:`~torch.cuda.graph_pool_handle` or
                 :meth:`other_Graph_instance.pool()<torch.cuda.CUDAGraph.pool>`) that hints this graph may share memory
                 with the indicated pool.  See :ref:`Graph memory management<graph-memory-management>`.
+            capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream.
+                Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
+                may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
+                actions in the current thread, and "relaxed" will not error on these actions.
```
Comment (albanD) [inline torch/cuda/graphs.py:74]: Not sure if this is 100% accurate for relaxed? But tbh the nvidia doc is not 100% clear either.
I would just link to the nvidia doc from there and say that our flags match these ones.

Same comment for the doc below

**Review by albanD — COMMENTED** [review]
cc @ptrblck can you confirm that this setting is ok to expose and we're not providing a big footgun to users here?

File: torch/csrc/cuda/Graph.cpp (hunk 4 above)
Comment (eellison) [reply torch/csrc/cuda/Graph.cpp:57]: This was the existing invocation: `torch::wrap_pybind_function_no_gil`

Comment (albanD) [reply torch/csrc/cuda/Graph.cpp:57]: But you use it around more code here.
If you're sure that absolutely no python code is used here, that's fine!

File: torch/csrc/cuda/Graph.cpp (hunk 3 above)
Comment (eellison) [reply torch/csrc/cuda/Graph.cpp:55]: I was trying to clean up this call: https://github.com/pytorch/pytorch/pull/107407/files#diff-39e542d87359e7d5381d036cbcea9ec759fbe469578bcdc5693ce6cfab7f1a54L72-L77.

Comment (albanD) [reply torch/csrc/cuda/Graph.cpp:55]: But you still pass None here. So this default value will never be used?
And I guess our custom typecast is already properly translating the `None` that you pass in python into a `nullopt` here.

File: torch/csrc/cuda/Graph.cpp (hunk 4 above)
Comment (eellison) [reply torch/csrc/cuda/Graph.cpp:57]: Sorry, how do I use it around more code here ? what's the difference ?

File: torch/csrc/cuda/Graph.cpp (hunk 3 above)
Comment (eellison) [reply torch/csrc/cuda/Graph.cpp]: Removed arg defaults.

File: torch/csrc/cuda/Graph.cpp (hunk 4 above)
Comment (albanD) [reply torch/csrc/cuda/Graph.cpp:55]: The only difference is that the old code was only releasing for the call to capture_begin.
now the arg parsing and the output conversion (if any) is also covered.
I think the return type is a c++ type so we should be good here.

File: torch/cuda/graphs.py [hunk 6]
```diff
@@ -68,13 +68,13 @@ def capture_begin(self, pool=None):
             pool (optional): Token (returned by :func:`~torch.cuda.graph_pool_handle` or
                 :meth:`other_Graph_instance.pool()<torch.cuda.CUDAGraph.pool>`) that hints this graph may share memory
                 with the indicated pool.  See :ref:`Graph memory management<graph-memory-management>`.
+            capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream.
+                Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
+                may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
+                actions in the current thread, and "relaxed" will not error on these actions. See docs on `cudaStreamCaptureMode`.
```
Comment (albanD) [inline torch/cuda/graphs.py:74]:
```suggestion
                actions in the current thread, and "relaxed" will not error on these actions. Do NOT change this setting unless you're very familiar with `cudaStreamCaptureMode <https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g9d0535d93a214cbf126835257b16ba85>`_.
```

File: torch/cuda/graphs.py [hunk 7]
```diff
@@ -139,6 +139,10 @@ class graph:
             may share memory from the specified pool. See :ref:`Graph memory management<graph-memory-management>`.
         stream (torch.cuda.Stream, optional): If supplied, will be set as the current stream in the context.
             If not supplied, ``graph`` sets its own internal side stream as the current stream in the context.
+        capture_error_mode (str, optional): specifies the cudaStreamCaptureMode for the graph capture stream.
+            Can be "global", "thread_local" or "relaxed". During cuda graph capture, some actions, such as cudaMalloc,
+            may be unsafe. "global" will error on actions in other threads, "thread_local" will only error for
+            actions in the current thread, and "relaxed" will not error on actions. See docs on `cudaStreamCaptureMode`.
```
Comment (albanD) [inline torch/cuda/graphs.py:145]:
```suggestion
            actions in the current thread, and "relaxed" will not error on these actions. Do NOT change this setting unless you're very familiar with `cudaStreamCaptureMode <https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g9d0535d93a214cbf126835257b16ba85>`_.
```

**Review by albanD — APPROVED** [review]
Small doc suggestion, SGTM otherwise!

**Comment by bottler** [comment]
Why call the option `capture_error_mode` rather than `capture_mode` to match the Nvidia API? If user is launching reliable GPU work on multiple threads, with no errors in any case, this option lets them choose whether to capture only that thread's work or all thread's work, right?

**Comment by albanD** [comment]
> If user is launching reliable GPU work on multiple threads, with no errors in any case, this option lets them choose whether to capture only that thread's work or all threads' work, right?

This is NOT the case, and that's why we should add an "Error" in the argument!
The only thing this flag does is to control if unsafe API calls (like cudaMalloc) done in other thread will crash or not.
The capture is always per-stream. And you will always be able to enqueue work on that stream from any thread you want (or only on the current one if you prefer).

**Review by eqy — APPROVED** [review]

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48656014).
